### PR TITLE
cicd: add basic precommit hooks

### PR DESCRIPTION
**Pull request by jsstevenson**
* Add precommit + some minimally controversial hooks. Add a note in README.
* Run hooks over existing codebase. There's a few instances of trailing whitespace and incorrect line endings.
* Run checks for formatting-related hooks in CI

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -47,7 +47,8 @@ venv/%:
 #=> develop: install package in develop mode
 .PHONY: develop setup
 develop setup:
-	pip install -e .[dev,extras,notebooks]
+	pip install -e .[dev,extras,notebooks]; \
+	pre-commit install
```
Comment (jsstevenson) [inline Makefile:51]: this seems to work for (and make sense to) me but I'm always wary of makefile environment construction soooo

File: submodules/README.md
Comment (korikuzma) [inline submodules/README.md:1]: We shouldn't run checks in submodules dir

**Review by korikuzma — CHANGES_REQUESTED** [review]
> Not sure if this should be mentioned explicitly in the readme?

This would be nice

File: .github/workflows/python-cqa.yml
[the remark this answers is not in the document]
Comment (jsstevenson) [reply .github/workflows/python-cqa.yml:1]: Yes -- was originally using it but it a) doesn't seem particularly necessary, this is like a 4 line action, b) it's maintenance-only, I'm not sure how quick they would be to upgrade dependencies. They suggest `pre-commit.ci lite` but I think this is also some extra heft that we don't necessarily need, I think workflows can be abstracted enough as it is.

File: submodules/README.md
Comment (jsstevenson) [reply submodules/README.md:1]: pre-commit doesn't run on anything in a submodule, unless I'm mistaken. The only affected file in that directory is `submodules/README.md`.

Comment (korikuzma) [reply submodules/README.md:1]: Ope, my bad. Only saw submodules dir when I was reviewing quickly

**Review by korikuzma — APPROVED** [review]

**Pull request merged**